src/systems/filecoin_mining/storage_mining/storage_miner_actor.go

```diff
@@ -1,21 +1,309 @@
+	// 1. A proof must be submitted after the postRandomness for this proving
+	// period is on chain
+	// if rt.ChainEpoch < sm.ProvingPeriodEnd - challengeTime {
+	//   panic("too early")
+	// }
+
+	// 2. A proof must be a valid snark proof with the correct public inputs
+	// 2.1 Get randomness from the chain at the right epoch
+	// postRandomness := rt.Randomness(postSubmission.Epoch, 0)
+	// 2.2 Generate the set of challenges
+	// challenges := GenerateChallengesForPoSt(r, keys(sm.Sectors))
+	// 2.3 Verify the PoSt Proof
+	// verifyPoSt(challenges, TODO)
 
-func (sma *StorageMinerActor_I) SubmitPoSt(postProof base_mining.PoStProof, sectorStateSets sector.SectorStateSets) {
 	panic("TODO")
+	return true
 }
 
-func (sma *StorageMinerActor_I) CommitSector(onChainInfo sealing.OnChainSealVerifyInfo) {
-	currSector, found := sma.Sectors()[onChainInfo.SectorNumber()]
-	if !found {
-		sma.Sectors()[onChainInfo.SectorNumber()] = []sealing.OnChainSealVerifyInfo{onChainInfo}
-	} else {
-		newSector := append(currSector, onChainInfo)
-		sma.Sectors()[onChainInfo.SectorNumber()] = newSector
+func (sm *StorageMinerActor_I) purgeSectorNumber(sectorNo sector.SectorNumber) {
+	delete(sm.Sectors(), sectorNo)
+	delete(sm.SectorStates(), sectorNo)
+	sm.ProvingSet_.Remove(sectorNo)
+}
+
+// Decision is to currently account for power based on sector
+// with at least one active deals and deals cannot be updated
+// an alternative proposal is to account for power based on active deals
+// an improvement proposal is to allow storage deal update in a sector
+
+// TODO: decide whether declared faults sectors should be
+// penalized in the same way as undeclared sectors and how
+
+// SubmitPoSt Workflow:
+// - Verify PoSt Submission
+// - Process ProvingSet.GetOnes()
+//   - State Transitions
+//     - Committed -> Active and credit power
+//     - Recovering -> Active and credit power
+//   - Process Active Sectors (pay miners)
+// - Process ProvingSet.GetZeros()
+//	   - increment FaultCount
+//     - clear Sector and slash pledge collateral if count >= CONSECUTIVE_FAULT_COUNT_LIMIT
+// - Process Expired Sectors (settle deals and return storage collateral to miners)
+//     - State Transition
+//       - Faulted / Recovering / Active / Committed -> Cleared
+//     - Remove SectorNumber from Sectors, SectorStates, ProvingSet
+func (sm *StorageMinerActor_I) SubmitPoSt(postSubmission poster.PoStSubmission) {
+	// Verify correct PoSt Submission
+	isPoStVerified := sm.verifyPoStSubmission(postSubmission)
+	if !isPoStVerified {
+		// no state transition, just error out and miner should submitPoSt again
+		// TODO: proper failure
+		panic("TODO")
+	}
+
+	// The proof is verified, process ProvingSet.GetOnes():
+	// ProvingSet.GetOnes() contains SectorCommitted, SectorActive, SectorRecovering
+	// ProvingSet itself does not store states, states are all stored in SectorStates
+	var powerUpdate uint64 = 0
+
+	for _, sectorNo := range sm.ProvingSet_.GetOnes() {
```

Comment:
this sounds very expensive, the previous code did not need to loop through every sector of a miner, but only through the expired ones and the faulty ones..

Reply:
yeah, let us think about it some more